### DISCUSSION TITLE
A way to use default arguments when calling a function

### DISCUSSION THREAD

**Discussion by charpurrr** [godotengine/godot-proposals#11057] [Scripting]
**Problem:**
I've experienced multiple cases where it would be preferable to skip setting a parameter when calling a function.

**Example:**
```gdscript
func _init(
	to: PackedScene,
	from_overlay: Overlay = Overlay.PLAIN,
	to_overlay: Overlay = Overlay.PLAIN,
	overlay_speed: float = 0.2,
	fake_delay: float = 0,
) -> void:
```

This is a function I created to instance nice looking scene transitions, and it has a few configurable options.
However, sometimes I would like to change the `fake_delay` while keeping the `overlay_speed` the same as the default value (in case it changes in the future.)

You might tell me to simply swap the `fake_delay` and `overlay_speed` around, but that creates the opposite issue where I would want to change `overlay_speed` but have to set `fake_delay` to a hard value.

Let's look at what I have to do if I would want to change `fake_delay` without changing `overlay_speed`:
```gdscript
SceneTransition.new(
	play_scene,
	SceneTransition.Overlay.PLAIN,
	SceneTransition.Overlay.PLAIN,
	0.2, # This is redundant, because the default value is already 0.2. 
	# If the default value were to change, all instances like this would have to be manually checked to fix this hardcoded value.
	1.0,
)
```

**Proposed Solution:**
Have a way to simply use the default value of an argument through a character.
I.e.:
```gdscript
SceneTransition.new(
	play_scene,
	SceneTransition.Overlay.PLAIN,
	SceneTransition.Overlay.PLAIN,
	*, # Special character that sets this parameter to its default value as seen in SceneTransition's init function.
	1.0,
)
```

**Comment by dalexeev** [comment]
See:
* #902
* #11077

This is problematic to implement, given that different functions are called differently. For example, GDScript functions support non-constant parameter initializers, so we cannot evaluate missing optional parameters in advance and patch the call arguments. And dynamic typing does not allow us to always know which function will be called.

**Comment by Mickeon** [comment]
I've thought about it in the past. I thought about using the `pass` keyword to symbolise the default argument.

I am aware that built-in functions in the API would heavily benefit from this, especially the server classes. As the API is currently, it would really be awesome. Generally speaking, though, functions with a huge amount of parameters are bad practice and should be discouraged. In an ideal world, we would be uniting these parameters in a [struct](https://github.com/godotengine/godot-proposals/issues/7329), but as of writing this it's not possible.

See also Dalexeev's response, implementation-wise.
